### PR TITLE
Copy task SourceFolders

### LINKED ISSUE #5881: Copy task should support copying a directory easily
Copy task doesn't make it easy to copy the entire directory, it forces you to mess with wildcards etc. and it's gnarly

We should consider adding a new SourceFolder parameter on the Copy task that, if specified, pre-populates the SourceFiles with the entire directory. Of course SourceFolder and SourcesFiles should be mutually exclusive.

### PR DESCRIPTION
**Pull request by jrdodds**
Fixes #5881

### Context
Implementation of proposed `ITaskItem[] SourceFolders` parameter.

### Changes Made
Within Copy.cs:
- Added `ITaskItem[] SourceFolders` parameter
- Modified `ValidateInputs()` to support the new parameter
- Modified `InitializeDestinationFiles()`
  - Previously this method iterated `SourceFiles` to create `DestinationFiles` items given the `DestinationFolder`.
  - The `Copy` task actually only operates on `SourceFiles` and `DestinationFiles`.
  - Extended the method to:
    - Use `FileMatcher::GetFiles()` for each directory in `SourceFolders`
    - Create items in both `SourceFiles` and `DestinationFiles` for each file found

Added additional unit tests in Copy_Tests.cs.

Added new error MSB3894 `Copy.IncompatibleParameters` for the case where `SourceFolders` and `DestinationFiles` are both present. This is very close to the existing MSB3022 `Copy.ExactlyOneTypeOfDestination` but it seemed prudent not to re-use MSB3022.

### Testing
Tested on Windows 11 and macOS 12

Ran full suite of unit tests from the command line

Ran some sample project files

### Notes
This implementation conforms to the proposed design with the exception that it does not copy empty directories.

This implementation leverages the existing `FileMatcher::GetFiles()` which recurses the directory tree using multiple threads. `GetFiles()` returns a list of files. Empty directories are not identified.

The use of `FileMatcher` could be replaced. The `Copy` task could implement its own support for recursing a directory tree. The wildcard matching logic in `FileMatcher` is not relevant for the `Copy` task.

### CODE REVIEW

**Comment by jrdodds** [comment]
Can the build be re-run?

**Comment by AR-May** [comment]
> Can the build be re-run? Or is it necessary to push a new commit?

You can write "/azp run" in the comment to the PR to re-run checks.

**Comment by rainersigwald** [comment]
Yeah, only committers have permission for `/azp run`--and committers should use the finer-grained rerun capabilities in AzDO or GitHub to rerun only failed jobs that failed for known, tracked-by-bugs issues.

File: src/Tasks.UnitTests/Copy_Tests.cs
Comment (AR-May) [inline src/Tasks.UnitTests/Copy_Tests.cs:128]: Shouldn't the task return false when there are no source folders or files? I am confused here.

**Review by AR-May — APPROVED** [review]
Overall looks good to me. Good tests coverage as well.

File: src/Tasks.UnitTests/Copy_Tests.cs
Comment (jrdodds) [reply src/Tasks.UnitTests/Copy_Tests.cs:128]: The existing code returns true when there are no `SourceFiles`.

https://github.com/dotnet/msbuild/blob/ed404c0d68dc073a09c5c9402a5c56f65e7a29a8/src/Tasks/Copy.cs#L400-L411

However, this test is never true when the task is used in MSBuild code because the `SourceFiles` parameter is `[Required]`. Because either or both `SourceFiles` and `SourceFolders` can be provided, neither can be `[Required]`.

To be more consistent with the existing behavior of the `Copy` task, I'll make the "empty set" an error.

**Comment by jrdodds** [comment]
Just a note that a unit test unrelated to the commit(s) failed and then on the next build passed.

**build:** Windows Core
**test run:** Microsoft.Build.Engine.UnitTests_net8.0_x64
**unit test:** Microsoft.Build.UnitTests.BackEnd.BuildManager_Tests.OutOfProcProjectInstanceBasedBuildDoesNotReloadFromDisk(shouldSerializeEntireState: True)

The log shows

```
error MSB4223: A node of the required type OutOfProc could not be created.
```

File: src/Tasks.UnitTests/Copy_Tests.cs
Comment (jrdodds) [reply src/Tasks.UnitTests/Copy_Tests.cs]: changes have been committed

Comment (jrdodds) [reply src/Tasks.UnitTests/Copy_Tests.cs]: `[Required]` doesn't distinguish between a null list and an empty list and doesn't allow an empty list. Failing on an empty list actually matches the current behavior of the task when used from MSBuild.

The behavior change is only when the task is used directly in code bypassing the `[Requireed]` logic.

Comment (jrdodds) [reply src/Tasks.UnitTests/Copy_Tests.cs]: > Sorry for my long absence, @jrdodds, I was sick. ...

Sorry to hear you were sick and I hope you have fully recovered.

Comment (jrdodds) [reply src/Tasks.UnitTests/Copy_Tests.cs:134]: Restored the current behavior. An empty set for sources is accepted and the task returns `true`.

File: src/Tasks.UnitTests/Copy_Tests.cs
Comment (AR-May) [inline src/Tasks.UnitTests/Copy_Tests.cs:135]: nit: Destination files are not checked anymore in this test. Is that intentional?

File: src/Tasks/Copy.cs
Comment (AR-May) [inline src/Tasks/Copy.cs:672]: I would prefer to do not have check `DestinationFiles.Length == 0` here: this situation is caught few lines below with more detailed error `TwoVectorsMustHaveSameLength`. I see similarities with the SourceFiles conditions, why, I suppose, this line was changed, but since it anyway would be some error, let's pick the more elaborate one.

**Review by AR-May — CHANGES_REQUESTED** [review]

File: src/Tasks/Copy.cs
Comment (jrdodds) [reply src/Tasks/Copy.cs]: Makes sense

**Review by MichalPavlik — APPROVED** [review]

**Comment by MichalPavlik** [comment]
LGTM Merge after Alina's comments are addressed.

**Comment by jrdodds** [comment]
Just pushed a commit to address both comments

**Review by AR-May — COMMENTED** [review]
LGTM

**Review by AR-May — APPROVED** [review]

**Comment by AR-May** [comment]
Our main currently cannot take PRs with new features. We will merge this as soon as our main is open.

**Review by MichalPavlik — APPROVED** [review]

**Comment by KirillOsenkov** [comment]
Is the new capability documented anywhere? I’d like to see a couple of examples.

**Comment by AR-May** [comment]
Feature is fully described in the issue with feature request:
https://github.com/dotnet/msbuild/issues/5881#issuecomment-1207271585
But it would be good idea to add documentation about it somewhere else too.

**Comment by KirillOsenkov** [comment]
does the description in the issue match the actual implementation? I’m afraid documenting a feature in an issue is insufficient. Ideally if there’s existing documentation for the Copy task, we should augment that.

**Comment by AR-May** [comment]
Yes, I agree, documenting in an issue is insufficient, we should add this to https://learn.microsoft.com/en-us/visualstudio/msbuild/copy-task.
Yes, the implementation matches the suggested design.

**Pull request merged**
